### ISSUE TITLE
Training metrics are being accumulated and not reset on train end by default

### ISSUE THREAD

**Issue by adeandrade** [PyTorchLightning/pytorch-lightning#11601]
## 🐛 Bug

Using `self.log` inside `training_step` accumulates the metrics, which are never reset before the validation step starts, resulting in validation results that correspond to the union of the training and validation datasets.

### To Reproduce

```python
class BoringModel(LightningModule):
    def __init__(
            self,
            size: Tuple[int, int],
            num_classes: int,
    ) -> None:

        super().__init__()

        self.save_hyperparameters(logger=False)

        self.size = size
        self.num_classes = num_classes

        ...

        self.segmentation_accuracy = Accuracy(num_classes=num_classes)
        self.segmentation_iou = JaccardIndex(num_classes)

    def calculate_loss(self, batch: BatchType) -> Dict[str, Tensor]:
        ...

        loss = loss_segmentation + loss_depth

        self.segmentation_accuracy(output_segmentation, segmentation)
        self.segmentation_iou(output_segmentation, segmentation)

        metrics = {
            'Loss': loss,
            'Loss Segmentation': loss_segmentation,
            'Loss Depth': loss_depth,
            'Segmentation Accuracy': self.segmentation_accuracy,
            'Segmentation IoU': self.segmentation_iou,
        }

        return metrics

    def training_step(self, batch: BatchType, *args) -> Tensor:
        metrics = self.calculate_loss(batch)

        for name, value in metrics.items():
            self.log(f'Train {name}', value)

        loss = metrics['Loss']

        return loss

    def validation_step(self, batch: BatchType, *args) -> None:
        metrics = self.calculate_loss(batch)

        for name, value in metrics.items():
            self.log(f'Validation {name}', value)

    def test_step(self, batch: BatchType, *args) -> None:
        metrics = self.calculate_loss(batch)

        for name, value in metrics.items():
            self.log(f'Test {name}', value)

    def on_validation_start(self) -> None:
        super().on_validation_start()

        num_samples1 = (
            self.segmentation_accuracy.tp +
            self.segmentation_accuracy.fp +
            self.segmentation_accuracy.tn +
            self.segmentation_accuracy.fn
        ) / self.size[0] / self.size[1] / self.num_classes
        num_samples2 = torch.sum(self.segmentation_iou.confmat) / self.size[0] / self.size[1]
        print('Validation start', num_samples1, num_samples2)

    def on_validation_end(self) -> None:
        num_samples1 = (
            self.segmentation_accuracy.tp +
            self.segmentation_accuracy.fp +
            self.segmentation_accuracy.tn +
            self.segmentation_accuracy.fn
        ) / self.size[0] / self.size[1] / self.num_classes
        num_samples2 = torch.sum(self.segmentation_iou.confmat) / self.size[0] / self.size[1]
        print('Validation end', num_samples1, num_samples2)

        super().on_validation_end()
```

Here's how I am training the model:

```python
    trainer = Trainer(
        max_epochs=max_num_epochs,
        logger=logger,
        enable_progress_bar=False,
        log_every_n_steps=utils.get_num_steps(dataset_train, batch_size),
        gpus=torch.cuda.device_count(),
        callbacks=[
            EarlyStopping(
                monitor='Validation Segmentation IoU',
                mode='max',
                patience=patience,
            ),
            ModelCheckpoint(
                monitor='Validation Segmentation IoU',
                mode='max',
            ),
        ],
    )

    trainer.fit(
        model=model,
        train_dataloaders=DataLoader(
            dataset=dataset_train,
            batch_size=batch_size,
            shuffle=True,
            num_workers=4,
        ),
        val_dataloaders=DataLoader(
            dataset=dataset_validation,
            batch_size=batch_size,
            num_workers=4,
        ),
    )
```

### Expected behavior

The training dataset has 2975 samples and the validation dataset has 500 samples. The validation metrics should be computed on 500 samples but I am getting 3475 (2975 + 500). Here's the output from running this code:

```
Validation start tensor(0., device='cuda:0') tensor(0., device='cuda:0')
Validation end tensor(20., device='cuda:0') tensor(20., device='cuda:0')
Validation start tensor(2975., device='cuda:0') tensor(2975., device='cuda:0')
Validation end tensor(3475., device='cuda:0') tensor(3475., device='cuda:0')
Validation start tensor(2975., device='cuda:0') tensor(2975., device='cuda:0')
Validation end tensor(3475., device='cuda:0') tensor(3475., device='cuda:0')
Validation start tensor(2975., device='cuda:0') tensor(2975., device='cuda:0')
Validation end tensor(3475., device='cuda:0') tensor(3475., device='cuda:0')
Validation start tensor(2975., device='cuda:0') tensor(2975., device='cuda:0')
Validation end tensor(3475., device='cuda:0') tensor(3475., device='cuda:0')
Validation start tensor(2975., device='cuda:0') tensor(2975., device='cuda:0')
Validation end tensor(3475., device='cuda:0') tensor(3475., device='cuda:0')
Validation start tensor(2975., device='cuda:0') tensor(2975., device='cuda:0')
Validation end tensor(3475., device='cuda:0') tensor(3475., device='cuda:0')
```

### Environment

- PyTorch Lightning Version: 1.5.9
- PyTorch Version: 1.10.1
- Python version: 3.9.6
- OS (e.g., Linux): Linux
- CUDA/cuDNN version: 11.3
- GPU models and configuration: Tested on diverse GPUs (V100, A100, Titan X).
- How you installed PyTorch (`conda`, `pip`, source): pip

### Additional context

Setting `on_epoch=False` when logging on `training_step` seems to not have any effect: the metrics are still aggregated and they do not reset before validation starts.

**Comment by rohitgr7** [comment]
hey @adeandrade !

you shouldn't use the same metric instance for both training and validation since they contain states which might differ for both.

check out the note here: https://torchmetrics.readthedocs.io/en/latest/pages/overview.html

**Comment by adeandrade** [comment]
> hey @adeandrade !
> 
> you shouldn't use the same metric instance for both training and validation since they contain states which might differ for both.
> 
> check out the note here: https://torchmetrics.readthedocs.io/en/latest/pages/overview.html

Thank you @rohitgr7. I'll use this approach then. Though It would simplify things if we could reuse metrics across stages. Why shouldn't we just reuse the metric and reset the state before validation starts? Is there a use case where the state needs to be kept across epochs?

**Comment by rohitgr7** [comment]
in Lightning, we support multiple validations within a single training epoch, so for consistency and in a general case, the validation loop runs right after the last training batch but the training loop isn't completed yet, thus the metrics associated with it aren't reset either when validation loop starts. This is not the only reason why it recommends using different instances but since you are using modular ones, it stores states internally, so ideally keeping them separated ensures they don't mix up and computes independently.
